### PR TITLE
[Kernel][Backend][Model] Blocksparse flash attention kernel and Phi-3-Small model

### PR DESCRIPTION
**Pull request by linxihui**
- Supports of Microsoft Phi-3-Small-8K and Phi-3-Small-128K models, which use blocksparse flash attention
- Prefilling Triton kernel for block-sparse attn
- Modified paged attention CUDA with the block-sparse attention, which allows hybrid sparsity pattern for each attention head.
- Use torch SPDA in prefilling phase for V100 or older GPUs, as well as CPU

This is joint work between Microsoft GenAI @linxihui, @beagleski, and vLLM @zhuohan123, @simon-mo @youkaichao.

### CODE REVIEW

File: vllm/attention/ops/blocksparse_attention/interface.py [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+        dtype=None,
+        homo_head=False,
+        active_head_range=None,
+        q_block_size=None,
+        use_spda=None,
+    ):
+        super().__init__()
+        if use_spda is None:
+            use_spda = is_hip() or is_cpu() or \
+                       torch.cuda.get_device_capability()[0] < 8
```
Comment (mgoin) [inline vllm/attention/ops/blocksparse_attention/interface.py:36]:
```suggestion
            use_spda = is_hip() or is_cpu() or not \
                       IS_COMPUTE_8_OR_ABOVE
```

File: vllm/model_executor/models/phi3small.py [hunk 2]
```diff
@@ -0,0 +1,448 @@
+import math
```
Comment (mgoin) [inline vllm/model_executor/models/phi3small.py:1]: Rename to `phi3_small.py` to match other models. Would it be more general to call this `phi3_blocksparse.py` if there are other sized Phi models with this attention?

File: vllm/model_executor/models/phi3small.py [hunk 3]
```diff
@@ -0,0 +1,448 @@
...
+from vllm.model_executor.layers.logits_processor import LogitsProcessor
+from vllm.model_executor.layers.quantization.base_config import (
+    QuantizationConfig)
+from vllm.model_executor.layers.rotary_embedding import get_rope
+from vllm.model_executor.layers.sampler import Sampler
+from vllm.model_executor.layers.vocab_parallel_embedding import (
+    DEFAULT_VOCAB_PADDING_SIZE, ParallelLMHead, VocabParallelEmbedding)
+from vllm.model_executor.model_loader.weight_utils import default_weight_loader
+# from vllm.model_executor.models.phi3small.phi3small_attention import (
+#     BlockSparseFlashAttention)
```
Comment (mgoin) [inline vllm/model_executor/models/phi3small.py:24]: cruft

File: vllm/model_executor/models/phi3small.py [hunk 4]
```diff
@@ -0,0 +1,448 @@
...
+    def weight_loader(self, param: torch.nn.Parameter,
+                      loaded_weight: torch.Tensor):
+        return load_column_parallel_weight(param, loaded_weight)
+
+
+class MergedColumnParallelLinear2(MergedColumnParallelLinear):
+
+    def weight_loader(self, param: torch.nn.Parameter,
+                      loaded_weight: torch.Tensor):
+        return load_column_parallel_weight(param, loaded_weight)
```
Comment (mgoin) [inline vllm/model_executor/models/phi3small.py:52]: Please give these more informative or explicitly overridden names - I initially missed the "2" when reading the model implementation.

File: vllm/entrypoints/openai/serving_engine.py [hunk 5]
```diff
@@ -100,6 +100,7 @@ def _create_logprobs(
                 token_logprob = step_top_logprobs[token_id].logprob
                 token = step_top_logprobs[token_id].decoded_token
                 logprobs.tokens.append(token)
+                token_logprob = max(token_logprob, -9999.0)
```
Comment (mgoin) [inline vllm/entrypoints/openai/serving_engine.py:103]: Is there any comment you can leave to explain why this specific number?

File: vllm/attention/ops/blocksparse_attention/utils.py [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+                    q_len,
+                    max_seqlen,
+                    dtype,
+                    device,
+                    block_size,
+                    local_blocks,
+                    vert_stride,
+                    return_dense,
+                ))
+            crow = crow[None].expand(n_heads, crow.shape[0])
```
Comment (mgoin) [inline vllm/attention/ops/blocksparse_attention/utils.py:178]: I've never seen `[None]` indexing used, what is this achieving?

File: vllm/attention/ops/blocksparse_attention/utils.py [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+
+
+def binary_mask_to_bias(mask_dense):
+    mask_dense = 1 - mask_dense
+    mask_dense.masked_fill_(mask_dense.bool(), -torch.inf)
+    return mask_dense
+
+
+def get_head_sliding_step(n_heads, vert_stride, homo_head=False):
+    # if vert_stride <= n_heads, rotating the heads
```
Comment (mgoin) [inline vllm/attention/ops/blocksparse_attention/utils.py:134]: Should we assert this?

File: vllm/attention/ops/blocksparse_attention/blocksparse_attention_kernel.py [hunk 8]
```diff
@@ -0,0 +1,422 @@
...
+        q = tl.load(
+            Q + offs_m[:, None] * stride_qt + offs_d[None, :] * stride_qd,
+            mask=(offs_m[:, None] < q_seqlen) & (offs_d[None, :] < D_HEAD),
+            other=0,
+        )
+
+    sparse_crow_ptr = (layout_crow_ptr + off_h * layout_crow_stride_h +
+                       q_pbid * layout_crow_stride_m)
+
+    # TODO(linxihui): load at once, supported in new Triton
```
Comment (mgoin) [inline vllm/attention/ops/blocksparse_attention/blocksparse_attention_kernel.py:332]: Can you link to a commit so we know which release we can switch to?

File: vllm/attention/backends/blocksparse_attn.py [hunk 9]
```diff
@@ -0,0 +1,339 @@
...
+        kv_caches: List[torch.Tensor],
+        src_to_dists: Dict[int, List[int]],
+    ) -> None:
+        PagedAttention.copy_blocks(kv_caches, src_to_dists)
+
+
+@dataclass
+class BlocksparseFlashAttentionMetadata(AttentionMetadata,
+                                        PagedAttentionMetadata):
+    """A Ccopy of Metadata for FlashAttentionBackend,
```
Comment (mgoin) [inline vllm/attention/backends/blocksparse_attn.py:126]:
```suggestion
    """A copy of Metadata for FlashAttentionBackend,
```

**Review by mgoin — COMMENTED** [review]
Thank you for your contribution, looking forward to this model release!

File: vllm/model_executor/models/phi3small.py (hunk 2 above)
Comment (linxihui) [reply vllm/model_executor/models/phi3small.py:1]: @mgoin  Good point, but indeed, only the phi3_small uses blocksparse attention.

File: vllm/model_executor/models/phi3small.py (hunk 4 above)
Comment (linxihui) [reply vllm/model_executor/models/phi3small.py:52]: @mgoin  sure, hard to give an expressive name. maybe call it `HeadMajorQKVParallelLinear`? as the heads change slower than the alternating of qkv>

File: vllm/entrypoints/openai/serving_engine.py (hunk 5 above)
Comment (linxihui) [reply vllm/entrypoints/openai/serving_engine.py:103]: @mgoin  This number already been in used in [line 111](https://github.com/vllm-project/vllm/blob/dfc07c7a43375504494bbdfd48a386263071b214/vllm/entrypoints/openai/serving_engine.py#L111) for case like `-inf`. I took from there.

File: vllm/attention/ops/blocksparse_attention/utils.py (hunk 6 above)
Comment (linxihui) [reply vllm/attention/ops/blocksparse_attention/utils.py:178]: `tensor[None]` is same as `tensor.unsqueeze(0)`, expanding a trivial dimension in both `torch` and `numpy`, pretty standard operation.

File: vllm/attention/ops/blocksparse_attention/blocksparse_attention_kernel.py (hunk 8 above)
Comment (linxihui) [reply vllm/attention/ops/blocksparse_attention/blocksparse_attention_kernel.py:332]: I know any Triton >=3.0.0 (E.g., 57e974dffc903a523de8ea6c5ad6eb16d561f83b) that support `tl.split` should work. But ya, I'll add to the comment.

Comment (linxihui) [reply vllm/attention/ops/blocksparse_attention/blocksparse_attention_kernel.py]: Comment updated.

**Comment by linxihui** [comment]
@mgoin  Thanks so much for reviewing. I've replied and made changed accordingly.

**Comment by beagleski** [comment]
@WoosukKwon Any suggestions on the kernel side?

**Comment by simon-mo** [comment]
I made a pass. I think once this PR adds unit test for both the Triton and PagedAttention kernels it should be good to go. You might also need to run clang-format to fix the merge conflict.

**Comment by linxihui** [comment]
> I made a pass. I think once this PR adds unit test for both the Triton and PagedAttention kernels it should be good to go. You might also need to run clang-format to fix the merge conflict.

Thanks @simon-mo for the review. I'll add the mising unitests today.

**Comment by simon-mo** [comment]
I have tested the PR locally as well.

**Pull request merged**